internal/api/server/apirequest_log_test.go

```diff
@@ -0,0 +1,180 @@
+				Endpoint: "/api/v03/courses/users/list",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					UserEmail: "server-admin@test.edulinq.org",
+					CourseID:  "course101",
+				},
+			},
+		},
+
+		// Valid permissions, APIRequestAssignmentContext request.
+		{
+			email:    "server-admin@test.edulinq.org",
+			endpoint: "courses/assignments/get",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/courses/assignments/get",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					UserEmail:    "server-admin@test.edulinq.org",
+					CourseID:     "course101",
+					AssignmentID: "hw0",
+				},
+			},
+		},
+
+		// Invalid permissions, APIRequestUserContext request.
+		{
+			email:           "course-student",
+			endpoint:        "users/list",
+			expectedLocator: "-041",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/users/list",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					UserEmail: "course-student@test.edulinq.org",
+				},
+				Locator: "-041",
+			},
+		},
+
+		// Invalid permissions, APIRequestCourseUserContext request.
+		{
+			email:           "course-student",
+			endpoint:        "courses/users/list",
+			expectedLocator: "-020",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/courses/users/list",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					UserEmail: "course-student@test.edulinq.org",
+					CourseID:  "course101",
+				},
+				Locator: "-020",
+			},
+		},
+
+		// Invalid permissions, APIRequestAssignmentContext request.
+		{
+			email:           "server-user",
+			endpoint:        "courses/assignments/get",
+			expectedLocator: "-040",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/courses/assignments/get",
+				Locator:  "-040",
+			},
+		},
+
+		// Valid permissions, invalid field request.
+		{
+			email:           "server-admin",
+			endpoint:        "courses/assignments/get",
+			fields:          map[string]any{"assignment-id": "zzz"},
+			expectedLocator: "-022",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/courses/assignments/get",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					CourseID:     "course101",
+					AssignmentID: "zzz",
+				},
+				Locator: "-022",
+			},
+		},
+	}
+
+	port, err := util.GetUnusedPort()
+	if err != nil {
+		test.Fatalf("Failed to get an unused port: '%v'.", err)
+	}
+
+	oldPort := config.WEB_HTTP_PORT.Get()
+	config.WEB_HTTP_PORT.Set(port)
+	defer config.WEB_HTTP_PORT.Set(oldPort)
+
+	// Adjust the server address for test client.
+	oldURL := core.SetTestServerURL(fmt.Sprintf("http://127.0.0.1:%d", port))
+	defer core.SetTestServerURL(oldURL)
+
+	for i, testCase := range testCases {
+		db.ResetForTesting()
+
+		runServerTestBaseFull(test, testCase.endpoint, testCase.fields, testCase.email, testCase.expectedLocator, fmt.Sprintf("Case %d: ", i))
```

Comment:
I don't think you will need to use this function (see above comment),
but if you do then you need to modify it so you can test test.Errorf() instead of test.Fatalf().
Otherwise you will be skipping test cases on each error.